### DISCUSSION TITLE
GUI for containerised control systems

### DISCUSSION THREAD

**Discussion by gilesknap** [epics-containers/epics-containers.github.io#5] [Ideas]
Ideally a full Web GUI will be available for EPICS at some point and be served out of a container that shares a pod with the IOC. But in the meantime is it possible to use a file based OPI such as Phoebus and still serve the files from a 2nd container? I feel it is important to place the GUI along side the IOC container so that version management is bundled up and no separate mechanism is required to tie IOC versions with their screens.

In the Containers meeting at ICALEPCS there was some discussion of running noVNC in a Kubernetes pod and opening a port to it via nodePort.

**Comment by thomascobb** [comment]
I think we will always want a site specific container for the screen build, so divergence on this would be fine, but I agree that putting them alongside the IOC container is important so versions match. Phoebus screens via an HTTP server also opens up the possibility of a rest API for screen, so you would point your action button at `<gui_container>/<device_pv_prefix>` or `<gui_container>/<device_pv_prefix>?screen=embedded` to get the screen source for a particular pv prefix without having to know the name of the screen file

**Reply by tynanford** [reply]
+1 for Phoebus and HTTP server

**Reply by thomascobb** [reply]
And we're already thinking of using https://github.com/tynanford/phoebusgen to generate screens on the fly...

**Comment by amichelotti** [comment]
> Ideally a full Web GUI will be available for EPICS at some point and be served out of a container that shares a pod with the IOC. But in the meantime is it possible to use a file based OPI such as Phoebus and still serve the files from a 2nd container? I feel it is important to place the GUI along side the IOC container so that version management is bundled up and no separate mechanism is required to tie IOC versions with their screens.
> 
> In the Containers meeting at ICALEPCS there was some discussion of running noVNC in a Kubernetes pod and opening a port to it via nodePort.

Maybe just an http ingress (or route if openshift) is needed to access to a novnc POD. In our k8s/cluster we provide pod ubuntu2204 framebuffered for noVNC with phoebus installed that mount opis. If you want to provide access x operator you can combine novnc/vnc with guacamore.

**Reply by gilesknap** [reply]
This sounds useful to me. We have gone with use of network=host for deploying IOCs to beamlines as there are many other clients to PVs than just the OPI. BUT this comes with its own security risks and is not available in our users personal K8S namespaces. This makes running a tutorial in an individual's namespace difficult. So I would definitely like to do this for our developer training.
